### DISCUSSION TITLE
How to configure static poco with static openssl but shared windows libraries on windows?

### DISCUSSION THREAD

**Discussion by norbertwie** [pocoproject/poco#4469] [Q&A]
In our Windows project we are using static poco and static openssl but the windows libraries in shared mode.

We use:
- https://github.com/pocoproject/poco.git with tag 1.13.1-release
- https://github.com/openssl/openssl version 3.2.1
- a slightly modified build script https://github.com/pocoproject/openssl/blob/develop/build.ps1 e.g. allowing version 3.x.

OpenSSL is compiled and build 
`-openssl_release 3.2.1 -vs_version 150 -config release -platform x64 -library static`

Poco is build with
BUILD_SHARED_LIBS=OFF
OPENSSL_USE_STATIC_LIBS=ON
OPENSSL_ROOT_DIR=\<path to openssl>

so far so good (one remark see below at the end).

But if we build now our project using "/MD" we got the link error libcrypt.lib not found (and libssl.lib).
After investigation I found in Crypto\include\Poco\Crypto\Crypto.h:

<pre>				#if !defined (_DLL)
					#define POCO_STATIC_SUFFIX "mt"
				#else // _DLL
					#define POCO_STATIC_SUFFIX ""
				#endif
</pre>
After a quick patch in the else part to

<pre>				#if !defined (_DLL)
					#define POCO_STATIC_SUFFIX "mt"
				#else // _DLL
					#define POCO_STATIC_SUFFIX "mt"
				#endif
</pre>

it worked.

The question is if we do something in general wrong or is this situation just not considered yet.
A solution might also be to use OPENSSL_USE_STATIC_LIBS instead of _DLL in the #if condition, but I don't have the overview of all OS flavors if this would work.
 
Remark:
Is the warning:
`External OpenSSL defined but internal headers used - possible mismatch!`
intended in our case? For me it sounds doing something wrong.

OffTopic:
Instead of "mt" I had to patch it to "_static" and also in the openssl build script as FindOpenSSl.cmake didn't found libs with "mt". But this is something I will investigate further.

**Comment by aleks-f** [comment]
> After a quick patch in the else part to
> 
> 				#if !defined (_DLL)
> 					#define POCO_STATIC_SUFFIX "mt"
> 				#else // _DLL
> 					#define POCO_STATIC_SUFFIX "mt"
> 				#endif
> it worked.

This is best addressed separately for the openssl libs - expecting them to follow the same binaries naming convention as we do obviously doesn't work.

> Remark: Is the warning: `External OpenSSL defined but internal headers used - possible mismatch!` intended in our case? For me it sounds doing something wrong.

this is a leftover from the times when we were building our own [windows openssl binaries](https://github.com/pocoproject/openssl); @matejk should be removed unless someone is interested to revive this (openssl build is a horrible nightmare, probably best to just rely on external distributions)

> OffTopic: Instead of "mt" I had to patch it to "_static" and also in the openssl build script as FindOpenSSl.cmake didn't found libs with "mt". But this is something I will investigate further.

same as above - looking for openssl libs should have its own logic, not relay on poco binary naming convention

**Reply by matejk** [reply]
Issue #4388 is partially related to removing custom OpenSSL build from Poco.

**Reply by norbertwie** [reply]
The issue is the #if condition, something like
<pre>
  		#if !defined (_DLL) || defined (OPENSSL_USE_STATIC_LIBS)
  			#define POCO_STATIC_SUFFIX "mt"
  		#else // _DLL
  			#define POCO_STATIC_SUFFIX ""
  		#endif
</pre>

might solve it.

OffTopic:
Naming convention might be ok. Openssl can produce names like libcryptmt.lib or libsslmt.lib for static libs. Here the issue is in FindOpenSSL.cmake provided by cmake which seems not to work - I will analyze this further.
